### ISSUE TITLE
Switch to LibPNG

### ISSUE THREAD

**Issue by faissaloo** [enigma-dev/enigma-dev#1391]
Is there any particular reason why we're using LodePNG? It's abandoned, slow and is often not on package managers. I'm currently trying to get LibEGM and Emake compiling on MacOS and this is one of the things making it difficult.

**Comment by RobertBColton** [comment]
Ok, I'll do it, from testing, libPNG is faster too and fundies has been asking me to do this so we don't have to check in lodepng in the repo (even though we do have it in shared). I've just been lazy about doing it, sorry. I will get to it!

**Comment by RobertBColton** [comment]
Alright, #1392 switches emake and libEGM to libpng. That just leaves LodePNG being used in the engine. Try that out for me and let me know if it makes the setup easier. I mean we already have our reasons for switching which you reiterated, but there's no rush until everybody is sure.

I don't know how performance will be different yet, I will do a benchmark in a second. I am actually doing the reading differently now by decoding a single row of the image at a time, so it may be faster or use a different amount of RAM.

**Comment by faissaloo** [comment]
libEGM compiles successfully now.

**Comment by RobertBColton** [comment]
Nice, huehue, I'll include the engine change in that pr too.

**Comment by RobertBColton** [comment]
Ok so I made some tweaks to emake in order to do a benchmark so we can prove to the whole wide hello world that we're doing this scientifically. My table below will compare the average time master and the pull request take, in microseconds, to load several projects five successive times.

```cpp
    } else if (ext == "gm81" || ext == "gmk" || ext == "gm6" || ext == "gmd") {
      buffers::Project* project;
      size_t runs = 5;
      double avgTime = 0.0;
      for (size_t i = 0; i < runs;) {
        auto start = std::chrono::high_resolution_clock::now();
        if (!(project = gmk::LoadGMK(input_file))) return 1;
        auto elapsed = std::chrono::high_resolution_clock::now() - start;
        long long microseconds = std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count();
        std::cout << "time " << i << " " << microseconds << " us" << std::endl;
        double delta = microseconds - avgTime;
        avgTime += delta/++i;
      }
      std::cout << "benchmark completed avgTime " << avgTime << " us" << std::endl;

      return 0;
```

(Give me a minute to collect all of the results in the markdown table generator)

**Comment by RobertBColton** [comment]
Alright, #1392 now switches the engine to libpng too. Only remnants of LodePNG at this point is in the makefiles, which I can do next. We can still use the shared sources to abstract libpng helpers so it's not as many additions. Overall, this looks like a pretty great change.

**Comment by RobertBColton** [comment]
Now that #1533 has been merged and we've finish the transition to Google Protocol Buffers, I am coming back and finally going to get this into master. Won't take me much effort to redo it, and this is the best time now we've eliminated a lot of the redundant places PNGs are handled.

**Comment by RobertBColton** [comment]
Going to redo the benchmark first so we can see how it shifted as a result of moving the protos directly inside the compiler.

|                         | Master/LodePNG (ms) | Pull Request/libpng (ms) |
|-------------------------|---------------------|--------------------------|
| fps6.gmk                |||
| Wild Racing.gmk         |||
| life.gm6                |||
| platform_engine.gmk     |||
| destructible_blocks.gmk |||

**Comment by RobertBColton** [comment]
Ok, once again, I decided to redo the pr a totally different way, and this time with great results. #1548 is not only faster it's also cleaner and more complete with LodePNG totally removed. I am now going to redo the benchmark one more time with the build times included.


|                         | Master/LodePNG (ms) | Pull Request/libpng (ms) |
|-------------------------|---------------------|--------------------------|
| fps6.gmk                |07381.8||
| Wild Racing.gmk         |16700.8||
| life.gm6                |06704.4||
| platform_engine.gmk     |07266.6||
| destructible_blocks.gmk |06436.6||

**Comment by RobertBColton** [comment]
Ok so after Josh pointed something out, I made another optimization in 9076bb3745e0fd4f32ed6c121181a20e81e9b905 that avoids additional allocations. This had another positive impact on the performance, so I am going to redo the benchmark again.

**Comment by RobertBColton** [comment]
Closing as resolved by #1548. Cheers!
